### PR TITLE
#2 Web-app: add filters by date and free rooms

### PR DESCRIPTION
**Pull request by Ekaterina1994**

### CODE REVIEW

File: gf-webapp/src/component/filter/AdultsFilter.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    if (isValueExist) {
+      setParams([...paramsWithoutAdults, [URL_QUERY_KEYS.Adults, adultsValueStringified]]);
+    } else {
+      setParams(paramsWithoutAdults);
+    }
+  }, [adultsValue]);
+
+  const minDate = new Date().toISOString().split("T")[0];
+  console.log(minDate);
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/AdultsFilter.tsx:45]: !

File: gf-webapp/src/component/filter/AdultsFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      .filter(param => param[0] !== URL_QUERY_KEYS.Adults);
+
+    if (isValueExist) {
+      setParams([...paramsWithoutAdults, [URL_QUERY_KEYS.Adults, adultsValueStringified]]);
+    } else {
+      setParams(paramsWithoutAdults);
+    }
+  }, [adultsValue]);
+
+  const minDate = new Date().toISOString().split("T")[0];
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/AdultsFilter.tsx:44]: Date formatter is commonly used

File: gf-webapp/src/component/filter/AdultsFilter.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      const adultsValueNumber: number = JSON.parse(adultsValueRaw);
+      setAdultsValue(adultsValueNumber);
+    }
+  }, []);
+
+  const onChangeInput = (event: React.ChangeEvent<InputType>) => {
+    setAdultsValue(Number(event.target.value));
+  };
+
+  const minAdultsValue = 0;
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/AdultsFilter.tsx:25]: Uppercase

Comment (tritonJS826) [reply gf-webapp/src/component/filter/AdultsFilter.tsx:25]: Why inside component

File: gf-webapp/src/component/filter/AdultsFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  const defaultAdultsValue: number = minAdultsValue;
+
+  useEffect(() => {
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/AdultsFilter.tsx:28]: Looks hard to understand

File: gf-webapp/src/component/filter/ChildrenFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  }, []);
+
+  const onChangeInput = (event: React.ChangeEvent<InputType>) => {
+    setChildrenValue(Number(event.target.value));
+  };
+
+  const minChildrenValue = 0;
+  const defaultChildrenValue: number = minChildrenValue;
+
+  useEffect(() => {
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/ChildrenFilter.tsx:28]: Looks like code duplication

File: gf-webapp/src/component/filter/Filter.module.scss [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+.label {
+  font-size: 24px;
+  color: $color-text;
+  margin-right: 30px;
+}
+
+.input {
+  margin: 10px 0 20px;
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/Filter.module.scss:22]: Old problem with constants
Create global

File: gf-webapp/src/component/filter/Filter.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+// type InputType = HTMLInputElement | HTMLTextAreaElement;
+
+// export enum URL_QUERY_KEYS {
+//   Search = "search",
+//   DateIn = "dateIn",
+//   DateOut = "dateOut",
+//   Adults = "adults",
+//   Children = "children",
+// }
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/Filter.tsx:21]: !

File: gf-webapp/src/component/filter/FilterContext.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  setSearchValue: STUB_FILTER_FUNCTION,
+  dateInValue: STUB_DATE_IN_VALUE,
+  setDateInValue: STUB_FILTER_FUNCTION,
+  dateOutValue: STUB_DATE_OUT_VALUE,
+  setDateOutValue: STUB_FILTER_FUNCTION,
+  adultsValue: STUB_ADULTS_VALUE,
+  setAdultsValue: STUB_FILTER_FUNCTION,
+  childrenValue: STUB_CHILDREN_VALUE,
+  setChildrenValue: STUB_FILTER_FUNCTION,
+} as FilterRooms);
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/FilterContext.ts:36]: Could we remove "as"

File: gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  };
+
+  useEffect(() => {
+    async function onRoomsInitialized() {
+      await initRooms();
+    }
+    onRoomsInitialized();
+  }, []);
+
+  const adultsFilterCondition = (roomFiltering: Room) => {
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx:24]: What mean room filtering? Do u mean filtered?

File: gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+      .toString()
+      .toLowerCase()
+      .includes(searchValue);
+  };
+
+  const dateFilterCondition = (roomFiltering: Room) => {
+    if (new Date(dateInValue).getTime() >= new Date(dateOutValue).getTime()
+      || new Date(dateInValue).getTime() < new Date().getTime()
+      || new Date(dateOutValue).getTime() < new Date().getTime()) {
+      alert("Wrong dates!!!");
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx:52]: Nice

File: gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+      return null;
+    } else if (dateInValue !== "yyyy-MM-dd" && dateOutValue !== "yyyy-MM-dd") {
+      const filteredRooms = roomFiltering.booked
+        .filter(room =>
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx:56]: Unreadable

File: gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        return filteredRooms;
+      }
+    } else {
+      return roomFiltering;
+    }
+  };
+
+  const filterCommonConditions = (roomsList: Room[]) => {
+    return roomsList
+      .filter((room) => adultsFilterCondition(room))
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx:72]: Why do u use addition arrow functions. Make it simplier, u don't need additolion functions

File: gf-webapp/src/component/filter/SearchFilter.tsx [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+
+export const SearchFilter = () => {
+  const {bookingRoomForm} = useDictionary().dictionary;
+  const [params, setParams] = useSearchParams();
+  const {searchValue, setSearchValue} = useFilterRooms();
+
+  useEffect(() => {
+    if (params.has(URL_QUERY_KEYS.Search)) {
+      setSearchValue(params.get(URL_QUERY_KEYS.Search) as string);
```
Comment (tritonJS826) [inline gf-webapp/src/component/filter/SearchFilter.tsx:15]: Avoid as

File: gf-webapp/src/logic/bookingPage/BookingPage.tsx [hunk 14]
```diff
@@ -1,17 +1,42 @@
...
 import {BookingPromo} from "src/logic/bookingPage/bookingPromo/BookingPromo";
-import {HotelPlan} from "src/logic/bookingPage/hotelPlan/HotelPlan";
+import {Filter} from "src/component/filter/Filter";
+import {FilterRoomsContext} from "src/component/filter/FilterContext";
+import {RoomsBlockWithFilters} from "src/component/filter/RoomsBlockWithFilters";
 import styles from "src/logic/bookingPage/BookingPage.module.scss";
 
 export const BookingPage = () => {
+  const [searchValue, setSearchValue] = useState("");
+  const [dateInValue, setDateInValue] = useState("yyyy-MM-dd");
```
Comment (tritonJS826) [inline gf-webapp/src/logic/bookingPage/BookingPage.tsx:11]: Magic string

File: gf-webapp/src/logic/bookingPage/BookingPage.tsx [hunk 15]
```diff
@@ -1,17 +1,42 @@
...
+  const [dateOutValue, setDateOutValue] = useState("yyyy-MM-dd");
+  const [adultsValue, setAdultsValue] = useState(0);
```
Comment (tritonJS826) [inline gf-webapp/src/logic/bookingPage/BookingPage.tsx:13]: Why 0?  Magic number

**Review by tritonJS826 — APPROVED** [review]

File: gf-webapp/src/component/filter/AdultsFilter.tsx (hunk 3 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/AdultsFilter.tsx:25]: Rename (ADULTS_VALUE_MIN) and move out

File: gf-webapp/src/component/filter/AdultsFilter.tsx (hunk 1 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/AdultsFilter.tsx:45]: delete unneeded console.log

File: gf-webapp/src/component/filter/AdultsFilter.tsx (hunk 2 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/AdultsFilter.tsx:44]: Moveout  const in utils, change function getCurrentDate to const by using methods .getDate, .getMonth, .getFullYear

File: gf-webapp/src/component/filter/ChildrenFilter.tsx (hunk 5 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/ChildrenFilter.tsx:28]: Delete all filters and create one filter component for all types of filtering

File: gf-webapp/src/component/filter/Filter.module.scss (hunk 6 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/Filter.module.scss:22]: Created global consts

Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/Filter.module.scss:22]: Created global consts

File: gf-webapp/src/component/filter/Filter.tsx (hunk 7 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/Filter.tsx]: Delete comments

File: gf-webapp/src/component/filter/FilterContext.ts (hunk 8 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/FilterContext.ts:36]: Remove

File: gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx (hunk 9 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx:24]: Change to room

File: gf-webapp/src/logic/bookingPage/BookingPage.tsx (hunk 15 above)
Comment (Ekaterina1994) [reply gf-webapp/src/logic/bookingPage/BookingPage.tsx]: Create const MIN_ADULTS_VALUE and MIN_CHILDREN_VALUE

File: gf-webapp/src/component/filter/AdultsFilter.tsx (hunk 4 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/AdultsFilter.tsx]: Move out function from useEffect and deleted duplicates

File: gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx (hunk 10 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx:52]: Create separate function to getTime from any date

File: gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx (hunk 11 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/RoomsBlockWithFilters.tsx:56]: Moveout function

File: gf-webapp/src/component/filter/SearchFilter.tsx (hunk 13 above)
Comment (Ekaterina1994) [reply gf-webapp/src/component/filter/SearchFilter.tsx]: For don't use as converted the next lines into
    const paramValueRaw = params.get(props.urlQueryKey);
      const paramValueNumber: number = paramValueRaw ? JSON.parse(paramValueRaw) : 0;
      const paramValueString: string = paramValueRaw ? paramValueRaw.slice(1, -1) : "";
    But i think it's not the best variant... Please help)

**Pull request merged**